### ISSUE TITLE
Passing weights to loss function

### ISSUE THREAD

**Issue by TKassis** [timeseriesAI/tsai#154]
I have a weighted loss function:

```python
class WMAELoss(nn.Module):
    def __init__(self, weights=None, reduction='mean'):
        super().__init__()
        assert reduction in ['mean', 'sum', 'none'], "You must set reduction to 'mean', 'sum' or 'none'"
        self.weights = weights
        
    def forward(self, input, target):
        loss = F.l1_loss(input.view(-1), target, reduction='none')
        if self.weights is not None:
            loss *= self.weights.expand_as(loss)
        if self.reduction == 'mean':
            return torch.mean(loss)
        elif self.reduction == 'sum':
            return torch.sum(loss)
        else:
            return loss
```
I would like to pass some calculated weights to it:

```python
weights = prepare_weights(y, 'sqrt_inv', lds=True)
```

I noticed that the TSDataLoaders.get_ts_dls accepts a weights argument such as

```python
dsets = TSDatasets(X, y, tfms=tfms, splits=splits, inplace=True)
dls = TSDataLoaders.from_dsets(dsets.train, dsets.valid, weights=weights)
```
But how do I pass the weights to the loss function? is there a dls.weights attribute or something similar that allows me to access the right weights for that batch?

Thanks,

**Comment by vrodriguezf** [comment]
If you want to pass instance weights you can use the [`WeightedPerSampleLoss` callback](https://timeseriesai.github.io/tsai/callback.core.html#WeightedPerSampleLoss) on top of any loss that you have.

**Comment by TKassis** [comment]
Thank you, this seems to be what I'm looking for. Do you know what is the 'weights' argument in the TSDataLoaders for then?

A related question, in WeightedPerSampleLoss the reduction seems to be always the sum. I'm trying to implement the Label Density Smoothing (LDS) from this: https://github.com/YyzHarry/imbalanced-regression but the author uses mean reduction in the weighted loss function. Do you know why one would take the sum vs. mean of the weighted losses?

**Comment by vrodriguezf** [comment]
IIRC, The `weights` argument, as in fastai's [WeightedDL](https://github.com/fastai/fastai/blob/5bc369c27846819edb825253fbb7bec0213cc2cb/fastai/callback/data.py#L17), has to do with the probability of the sample being part of a batch. It's used in the method `get_idxs` of the dataloader. I normally use that to deal with class imbalance, giving more weight to samples of the minority class to ensure their presence in every batch during training.

As for the use of sum reduction, I don't really know, I always use mean reduction because it's not affected by the batch size.

**Comment by TKassis** [comment]
Got it, thanks!
